### ISSUE TITLE
Checking permissions and roles with only ID, and not string 'name'

### ISSUE THREAD

**Issue by pkeogan** [spatie/laravel-permission#641]
Is there a way to do this? 

I find my self having to call the following code inside a Permission to, or HasRole

`Permission::find($id)->name` 

inside a call boolean call, like the following

`if(Auth::user()->hasPermissionTo(Permission::find($id)->name))) { 
... Do Something
}`

I find having to compare strings rather than ints is rather non-optimized for large scale executions.

Thoughts? am I simply reading the Documentation wrong?

**Comment by drbyte** [comment]
You're right: the `hasPermissionTo()` doesn't accept `int`s.  But it does accept `Permission` objects or names.
It sounds like something in your app has built yet another relation to permission IDs beyond the relationships in this package.
Can you explain or show how your app is geared solely around IDs instead of permission names/objects?

**Comment by pkeogan** [comment]
I sense I am possibly integrating the use of this lovely package wrong, so I apologize, if after you reading this reply you think to yourself "what is this guy thinking...".

In my application, I am restricting the ability to CRUD on dynamically created objects.

For example, I have a **Report** model. This model has an associated  the basic CRUD functions. Upon the construct of this object, permissions from this package are dynamical created then associated to this object.

Here is what the **Report** model migration would look like
```
// In Report Migration File
...
            $table->integer('view_report_permission')->nullable();
            $table->integer('create_report_permission')->nullable();
            $table->integer('view_deleted_reports_permission')->nullable();
            $table->integer('edit_report_permission')->nullable();
            $table->integer('delete_report_permission')->nullable();
            $table->integer('restore_report_permission')->nullable();
            $table->integer('permanently_delete_report_permission')->nullable();
            $table->softDeletes();
...

```

Upon construct of the  **Report** model, these permissions would be created, then saved to the model.
_foreach CRUD ability_
```
$report->slug = 'foo';
...
   //Created a Permission, with a name of 'view_report_foo', and id of '100'
  Permission::create(['name' => 'view_report_' . $report->slug]);

  //Find the permission, with the name of  'view_report_foo', then gets it's ID (100), then save that to the report
  $report->view_report_permission = Permission::all()->firstWhere('name', 'view_report_' . $report->slug)->id;
...
   //This would end up with the following on the Report model
   $report->slug == 'foo';
   $report->view_report_permission == '100'
...
```

Further on in the program, when needing to view said report, with a slug of 'foo', I could call a method in a blade.php file

```
// In a blade.php View
$report = Report::firstWhere('slug', 'foo');
...
   @permission($report->view_report_permission) // 
        // Do Something
   @endpermission
...
```

The custom blade directive would be from here
```
// In a custom Service Provider
...
    /**
     * Check if current User has the Permission ID Supplied, If user does, return true, if not, return false.
     *
     * @param Int $permissionID
     *
     * @return boolean
     */
     Blade::directive('permission', function ($permissionID) { return "<?php  if(Auth::user()->hasPermissionTo(App\Models\Auth\Permission::find(" . $permissionID . ")->name)): ?>"; });
      
    /**
     * Return the endif for a @permission Blade Directive
     *
     * @return void
     */
     Blade::directive('endpermission', function () { return '<?php endif; ?>'; });
      
...
```

So in long, that is how my app is geared to solely using ID's instead of permissions names/objects, and exactly how its being implemented.

**Comment by pkeogan** [comment]
Would it be better/more optimized if I used `Permission` Objects?

For example
```
//In Report Relationships Trait
...
    public function viewPermission()
    {
        return $this->belongsTo(Permission:class);
    }
...
```

Then I could make the following call
```
...
    if(Auth::user()->hasPermissionTo($this->viewPermission)) // Not sure if $this-> is a valid replacement for (Auth::user()->, but you can get the point
    {
        //do something
    }
...
```

**Comment by drbyte** [comment]
Thanks for sharing the details - that context is helpful!

Is the Permission ID number ever used anywhere else? 
I'm wondering whether you need to store the ID number anywhere at all, since it's always based on the `'permission_name_' . $report->slug`. 
If you do use it elsewhere, that's fine. But if not then maybe it's over-engineered? (been there, done that! :) )

I suppose there could be `n+1` considerations, however, given that the Permission data is already cached it may already be equally performant.
If the `int` is indeed needed, then, as per your last comment, you could explore keeping keep your `int` storage and connecting it using a `belongsTo` relationship.  I haven't found a need to do that myself yet, so am just trying to follow along with your thinking.

**Comment by pkeogan** [comment]
> Is the Permission ID number ever used anywhere else?
> I'm wondering whether you need to store the ID number anywhere at all, since it's always based on the 'permission_name_' . $report->slug.
> If you do use it elsewhere, that's fine. But if not then maybe it's over-engineered? (been there, done that! :) )

At one point, I did have the same thought, but I came to this conclusion;, in order to create a even more robust and dynamic system, I wanted to be able to edit a model's CRUD Permissions. So if for some reason during production, you wanted change a model's crud permissions you could. Sorry if that is hard to follow. 

> If the int is indeed needed, then, as per your last comment, you could explore keeping keep your int storage and connecting it using a belongsTo relationship. I haven't found a need to do that myself yet, so am just trying to follow along with your thinking.


I believe I will continue to explore the use of IDs of permissions with this package, and then submit a pull request when I develop something worthwhile and stable. I appreciate all of your input greatly!

**Comment by drbyte** [comment]
Thumbs up 👍
